### PR TITLE
test(e2e): permit2 approval

### PR DESCRIPTION
**Pull request by cartcrom**
## Description
adds tests for different cases of the token approval / permit2 approval flow

### CODE REVIEW

File: cypress/e2e/permit2.test.ts [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    cy.hardhat().then((hardhat) => {
+      const tokenApprovalStub = cy.stub(hardhat.wallet, 'sendTransaction')
+      tokenApprovalStub.rejects(USER_REJECTION) // reject token approval
+
+      cy.get('[data-testid="swap-approve-button"]').click()
+      cy.get('[data-testid="swap-approve-button"]')
+        .should('have.text', 'Approve use of DAI')
+        .then(() => {
+          const permitApprovalStub = cy.stub(hardhat.provider, 'send')
+          permitApprovalStub.withArgs('eth_signTypedData_v4').rejects(USER_REJECTION) // reject permit approval
```
Comment (cartcrom) [inline cypress/e2e/permit2.test.ts:128]: I had to nest this stub to get it to occur _after_ token balances are fetched from the provider, because for some reason `cy.stub(hardhat.provider,'send').withArgs('eth_signTypedData_v4').rejects(USER_REJECTION)` is also stubbing/failing send calls for token balances. 

Ideally I'd like to move `permitApprovalStub` to the top with `tokenApprovalStub`, but I'm not sure why the balance issue is happening yet.

Comment (earuak) [reply cypress/e2e/permit2.test.ts:128]: As I understand it, you are facing an issue handling sending calls to token balances and rejecting calls with the **eth_signTypedData_v4.** It looks like you had to nest a stub for it to occur after fetching the token balances from the provider, and now you're trying to figure out why you're still having a balance issue.

Without more context about the code and settings, it's difficult to give an exact answer to your problem. However, based on what has been shared, here are some suggestions that may help you resolve the issue:

Make sure the provider is configured correctly and calls to fetch token balances are being made correctly. If there is a problem with fetching balances, it can affect the balance.

Check that the stubs are being configured correctly. Make sure that the stubs' arguments and returns are correctly configured for the calls being made.

Try moving the **tokenApprovalStub** and **permitApprovalStub** to the top and see if that fixes the issue. If the problem persists, it may be necessary to revisit the code and see if there are any issues with the calls being made.

In short, it's difficult to determine the exact solution without more context and information about the code. However, make sure the provider is configured correctly, make sure the stubs are configured correctly, and try moving them to the top. Hope this helps.

File: cypress/e2e/permit2.test.ts [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        promises.push(hardhat.approval.setTokenAllowanceForPermit2({ owner, token: INPUT_TOKEN }))
+      }
+      if (type === 'permit2' || type === 'both') {
+        promises.push(hardhat.approval.setPermit2Allowance({ owner, token: INPUT_TOKEN }))
+      }
+
+      await Promise.all(promises)
+      return hardhat
+    })
+  }
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:55]: Don't use a shorthand util **with branching logic** in tests. You should inline this in the call sites.

File: cypress/e2e/permit2.test.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+  const expectTokenMaxApproval = () => {
+    // check token approval
+    return cy
+      .hardhat()
+      .then((hardhat) => hardhat.approval.getTokenAllowanceForPermit2({ owner, token: INPUT_TOKEN }))
+      .should('deep.equal', MaxUint256)
+  }
+
+  // Shorthand util for asserting permit2 approval for spend of the input token on-chain
+  const expectPermit2MaxApproval = (approvalTime: number) => {
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:67]: name it more verbosely (see above)

**Review by zzmp — CHANGES_REQUESTED** [review]

File: cypress/e2e/permit2.test.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+const APPROVE_BUTTON = '[data-testid="swap-approve-button"]'
+
+// The same tokens & swap-amount combination is used for all permit2 tests
+const INPUT_TOKEN = DAI
+const OUTPUT_TOKEN = USDC_MAINNET
+const TEST_BALANCE_INCREMENT = 0.01
+
+const setupSwap = () => {
+  // Use input/output in search params for faster test setup
+  cy.visit(`/swap/?inputCurrency=${INPUT_TOKEN.address}&outputCurrency=${OUTPUT_TOKEN.address}`, {
```
[the remark this answers is not in the document]
Comment (cartcrom) [reply cypress/e2e/permit2.test.ts:14]: I disagree. Since a swap is setup before every permit2 test, the fastest setup method should be used (url params). `swap.test.ts` should ensure there is no difference between manual vs url param swap setup, so `permit2.test.ts`'s usage of one vs the other is arbitrary, other than speed

File: cypress/e2e/permit2.test.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  // Completes the swap
+  cy.get('#swap-button').click()
+  cy.get('#confirm-swap-or-send').click()
+  cy.get('[data-testid="dismiss-tx-confirmation"]').click()
+
+  // There should be a successful Swap notification
+  cy.contains('Swapped').should('exist')
+}
+
+// TODO: Update tests to differentiate between permit2 vs token approval button text once UI is updated to indicate approval step
```
[the remark this answers is not in the document]
Comment (cartcrom) [reply cypress/e2e/permit2.test.ts:34]: checking w/ eddie if there's an existing ticket for that UI change

File: cypress/e2e/permit2.test.ts [hunk 6]
```diff
@@ -0,0 +1,206 @@
+import { MaxUint160, MaxUint256 } from '@uniswap/permit2-sdk'
+
+import { DAI, USDC_MAINNET } from '../../src/constants/tokens'
+
+const APPROVE_BUTTON = '[data-testid="swap-approve-button"]'
+
+/** Initiates a swap and confirms its success. */
+const swaps = () => {
+  // Ensures that the swap button is enabled, since it can be temporarily disabled following approval.
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:9]: nit: This comment is iffy because we don't know about the "approval" context from within this helper function. Could you make it more explicit?

File: cypress/e2e/permit2.test.ts [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+  // Completes the swap.
+  cy.get('#swap-button').click()
+  cy.get('#confirm-swap-or-send').click()
+  cy.get('[data-testid="dismiss-tx-confirmation"]').click()
+
+  // Verifies that there is a successful swap notification.
+  cy.contains('Swapped').should('exist')
+}
+
+// TODO: Update tests to differentiate between permit2 vs token approval button text once UI is updated to indicate approval step.
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:21]: Any TODO needs a linked jira ticket.
If there's already a ticket for the button text disambiguation, can you re-use that ticket number?

File: cypress/e2e/permit2.test.ts [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+describe('Permit2', () => {
+  // The same tokens & swap-amount combination is used for all permit2 tests.
+  const INPUT_TOKEN = DAI
+  const OUTPUT_TOKEN = USDC_MAINNET
+  const TEST_BALANCE_INCREMENT = 0.01
+
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:27]: NB: It's fine to specify them like this - I think it's _also_ fine to inline these values.
I think this is different than for implementations - in a test file I expect more inlining and explicit behavior.

File: cypress/e2e/permit2.test.ts [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+  beforeEach(() => {
+    // Uses input/output in search params for faster test setup.
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:29]:
```suggestion
    // Sets up a swap between INPUT_TOKEN and OUTPUT_TOKEN.
```

File: cypress/e2e/permit2.test.ts [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+  /** Asserts the universal router has a max permit2 approval for spend of the input token on-chain. */
+  const expectPermit2AllowanceForUniversalRouterToBeMax = (approvalTime: number) => {
+    return cy
+      .hardhat()
+      .then((hardhat) => hardhat.approval.getPermit2Allowance({ owner: hardhat.wallet, token: INPUT_TOKEN }))
+      .then((allowance) => {
+        cy.then(() => MaxUint160.eq(allowance.amount)).should('eq', true)
+        const expected = Math.floor((approvalTime + 2_592_000_000) / 1000)
+
+        // Asserts that the on-chain expiration is in 30 days, within a tolerance of 20 seconds.
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:54]:
```suggestion

        // Asserts that the on-chain expiration is in 30 days, within a tolerance of 20 seconds.
        const expected = Math.floor((approvalTime + 2_592_000_000) / 1000)
```

File: cypress/e2e/permit2.test.ts [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+      .should('deep.equal', MaxUint256)
+  }
+
+  /** Asserts the universal router has a max permit2 approval for spend of the input token on-chain. */
+  const expectPermit2AllowanceForUniversalRouterToBeMax = (approvalTime: number) => {
+    return cy
+      .hardhat()
+      .then((hardhat) => hardhat.approval.getPermit2Allowance({ owner: hardhat.wallet, token: INPUT_TOKEN }))
+      .then((allowance) => {
+        cy.then(() => MaxUint160.eq(allowance.amount)).should('eq', true)
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:51]: You don't need to use then, because you're not working with a promise:

```suggestion
        cy.wrap(MaxUint160.eq(allowance.amount)).should('eq', true)
```

File: cypress/e2e/permit2.test.ts [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        const expected = Math.floor((approvalTime + 2_592_000_000) / 1000)
+
+        // Asserts that the on-chain expiration is in 30 days, within a tolerance of 20 seconds.
+        cy.then(() => Math.abs(allowance.expiration - expected)).should('be.lessThan', 20)
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:55]: Use [`closeTo`](https://docs.cypress.io/guides/references/assertions#:~:text=closeTo(expected%2C%20delta)):

```suggestion
        cy.wrap(allowance.expiration).should('be.closeTo', expected, 20)
```

File: cypress/e2e/permit2.test.ts [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+      })
+  }
+
+  it('does not prompt approval when user has already approved token and permit2', () => {
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:59]: technically this isn't what you're testing:

```suggestion
  it('swaps when user has already approved token and permit2', () => {
```

File: cypress/e2e/permit2.test.ts [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+    cy.hardhat()
+      .then(({ approval, wallet }) => {
+        approval.setTokenAllowanceForPermit2({ owner: wallet, token: INPUT_TOKEN })
+        approval.setPermit2Allowance({ owner: wallet, token: INPUT_TOKEN })
+      })
+      .then(swaps)
+  })
+
+  it('can swap after completing full permit2 approval process', () => {
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:68]: nit: I prefer every test to start with a transitive verb:

```suggestion
  it('swaps after completing full permit2 approval process', () => {
```

File: cypress/e2e/permit2.test.ts [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+    cy.get(APPROVE_BUTTON).click()
+    const approvalTime = Date.now()
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:70]: This may not occur at the same time. Cypress queues up actions, but this is executed synchronously.

```suggestion
    cy.get(APPROVE_BUTTON).click().then(() => {
      const approvalTime = Date.now()
      ... // <- put the rest of the test in this closure, so that you have access to approvalTime.
    })
```

File: cypress/e2e/permit2.test.ts [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+
+      // Clicking the approve button should trigger a token approval that will be rejected by the user (tokenApprovalStub).
+      cy.get(APPROVE_BUTTON).click()
+
+      // The swap component should prompt approval again.
+      cy.get(APPROVE_BUTTON)
+        .should('have.text', `Approve use of ${INPUT_TOKEN.symbol}`)
+        .then(() => {
+          const permitApprovalStub = cy.stub(hardhat.provider, 'send')
+          permitApprovalStub.withArgs('eth_signTypedData_v4').rejects(USER_REJECTION) // reject permit approval
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:96]:
```suggestion
          permitApprovalStub.withArgs('eth_signTypedData_v4').rejects(USER_REJECTION) // reject permit approval
          permitApprovalStub.callThrough() // <- this tells it to just call the orig. impl. if the args don't match, otherwise it will stub *everything*
```

File: cypress/e2e/permit2.test.ts [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+        })
+    })
+  })
+
+  it('can swap with existing token approval and missing permit approval', () => {
+    cy.hardhat()
+      .then(({ approval, wallet }) => approval.setTokenAllowanceForPermit2({ owner: wallet, token: INPUT_TOKEN }))
+      .then(() => {
+        cy.get(APPROVE_BUTTON).click()
+        const approvalTime = Date.now()
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:129]: Wrap in `then`

File: cypress/e2e/permit2.test.ts [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+          // The user is now allowing approval, but the permit2 signature will be rejected by the user (permitApprovalStub).
+          cy.get(APPROVE_BUTTON).click()
+          cy.get(APPROVE_BUTTON)
+            .should('have.text', `Approve use of ${INPUT_TOKEN.symbol}`)
+            .then(() => {
+              permitApprovalStub.restore() // allow permit approval
+
+              // The swap should now be able to proceed, as the permit2 signature will be accepted by the user.
+              cy.get(APPROVE_BUTTON).click()
+              const approvalTime = Date.now()
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:109]: Wrap in `then`

File: cypress/e2e/permit2.test.ts [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+    const expiredAllowance = { expiration: Math.floor((Date.now() - 1) / 1000) }
+
+    cy.hardhat()
+      .then(({ approval, wallet }) => {
+        approval.setTokenAllowanceForPermit2({ owner: wallet, token: INPUT_TOKEN })
+        approval.setPermit2Allowance({ owner: wallet, token: INPUT_TOKEN }, expiredAllowance)
+      })
+      .then(() => {
+        cy.get(APPROVE_BUTTON).click()
+        const approvalTime = Date.now()
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:163]: Wrap in `then`

**Review by zzmp — CHANGES_REQUESTED** [review]

File: cypress/e2e/permit2.test.ts [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+
+/** Initiates a swap and confirms its success. */
+const swaps = () => {
+  // Ensures that the swap button is enabled, since it can be temporarily disabled following approval.
+  cy.get('#swap-button').should('not.have.attr', 'disabled')
+
+  // Completes the swap.
+  cy.get('#swap-button').click()
+  cy.get('#confirm-swap-or-send').click()
+  cy.get('[data-testid="dismiss-tx-confirmation"]').click()
```
Comment (JFrankfurt) [inline cypress/e2e/permit2.test.ts:15]: use testid util

File: cypress/e2e/permit2.test.ts [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+
+  beforeEach(() => {
+    // Uses input/output in search params for faster test setup.
+    cy.visit(`/swap/?inputCurrency=${INPUT_TOKEN.address}&outputCurrency=${OUTPUT_TOKEN.address}`, {
+      ethereum: 'hardhat',
+    })
+    cy.get('#swap-currency-input .token-amount-input').clear().type(TEST_BALANCE_INCREMENT.toString())
+  })
+
+  /** Asserts permit2 has a max approval for spend of the input token on-chain. */
```
Comment (JFrankfurt) [inline cypress/e2e/permit2.test.ts:36]:
```suggestion
  // Asserts permit2 has a max approval for spend of the input token on-chain.
```

File: cypress/e2e/permit2.test.ts [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+  const expectTokenAllowanceForPermit2ToBeMax = () => {
+    // check token approval
+    return cy
+      .hardhat()
+      .then(({ approval, wallet }) => approval.getTokenAllowanceForPermit2({ owner: wallet, token: INPUT_TOKEN }))
+      .should('deep.equal', MaxUint256)
+  }
+
+  /** Asserts the universal router has a max permit2 approval for spend of the input token on-chain. */
```
Comment (JFrankfurt) [inline cypress/e2e/permit2.test.ts:45]:
```suggestion
  // Asserts the universal router has a max permit2 approval for spend of the input token on-chain.
```

File: cypress/e2e/permit2.test.ts (hunk 11 above)
Comment (JFrankfurt) [reply cypress/e2e/permit2.test.ts:51]: You could also use `expect` here

File: cypress/e2e/permit2.test.ts [hunk 23]
```diff
@@ -0,0 +1,206 @@
+import { MaxUint160, MaxUint256 } from '@uniswap/permit2-sdk'
+
+import { DAI, USDC_MAINNET } from '../../src/constants/tokens'
+
+const APPROVE_BUTTON = '[data-testid="swap-approve-button"]'
```
[the remark this answers is not in the document]
Comment (cartcrom) [reply cypress/e2e/permit2.test.ts:5]: I think cmd-replacing `cy.get(APPROVE_BUTTON)` with `cy.getByTestId('swap-approve-button')` would be the same refactor effort as cmd-replacing `cy.get(getTestSelector('swap-approve-button'))`

File: cypress/e2e/permit2.test.ts (hunk 7 above)
Comment (cartcrom) [reply cypress/e2e/permit2.test.ts:21]: yes was waiting to see if there was an existing ticket but there wasn't so I made one

File: cypress/e2e/permit2.test.ts (hunk 15 above)
Comment (cartcrom) [reply cypress/e2e/permit2.test.ts:70]: I was wondering about that, thank you

File: cypress/e2e/permit2.test.ts (hunk 16 above)
Comment (cartcrom) [reply cypress/e2e/permit2.test.ts:96]: thank you! much better than the answer the bot replied to my PR with above

File: cypress/e2e/permit2.test.ts (hunk 22 above)
Comment (zzmp) [reply cypress/e2e/permit2.test.ts:47]: This should be a doc comment, not an implementation comment. It's documenting the function, not explaining the implementation.

File: cypress/e2e/permit2.test.ts (hunk 21 above)
Comment (zzmp) [reply cypress/e2e/permit2.test.ts:38]: This should be a doc comment, not an implementation comment. It's documenting the function, not explaining the implementation.

File: cypress/e2e/permit2.test.ts [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+    // Sets up a swap between INPUT_TOKEN and OUTPUT_TOKEN.
+    cy.visit(`/swap/?inputCurrency=${INPUT_TOKEN.address}&outputCurrency=${OUTPUT_TOKEN.address}`, {
+      ethereum: 'hardhat',
+    })
+    cy.get('#swap-currency-input .token-amount-input').clear().type(TEST_BALANCE_INCREMENT.toString())
+  })
+
+  /** Asserts permit2 has a max approval for spend of the input token on-chain. */
+  function expectTokenAllowanceForPermit2ToBeMax() {
+    // check token approval
```
Comment (zzmp) [inline cypress/e2e/permit2.test.ts:40]: nit:
```suggestion
```

**Review by zzmp — APPROVED** [review]

File: cypress/e2e/permit2.test.ts [hunk 25]
```diff
@@ -0,0 +1,223 @@
+import { MaxUint160, MaxUint256 } from '@uniswap/permit2-sdk'
+
+import { DAI, USDC_MAINNET } from '../../src/constants/tokens'
+import { getTestSelector } from '../utils'
+
+const APPROVE_BUTTON = '[data-testid="swap-approve-button"]'
+
+/** Initiates a swap and confirms its success. */
```
Comment (JFrankfurt) [inline cypress/e2e/permit2.test.ts:8]:
```suggestion
// Initiates a swap and confirms its success.
```

**Review by JFrankfurt — APPROVED** [review]

**Review by TDG510 — COMMENTED** [review]
Roderick Watson
